<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Oscar E. Ospina, Alex C. Soupir, Roberto Manjarres-Betancur, Guillermo Gonzalez-Calderon, Xiaoqing Yu, Brooke L. Fridley" />


<title>Using the STdiff function from the spatialGE package</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Spatial linear models for differential expression analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="diff_expr_spatial_linear_models.html">Manuscript figures</a>
</li>
<li>
  <a href="diff_expr_example_stdiff.html">Example using STdiff</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/oospina/diff_expression_spatial_linear_models">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Using the STdiff function from the
spatialGE package</h1>
<h4 class="author">Oscar E. Ospina, Alex C. Soupir, Roberto
Manjarres-Betancur, Guillermo Gonzalez-Calderon, Xiaoqing Yu, Brooke L.
Fridley</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-06-21
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong>
<code>diff_expression_spatial_linear_models/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240118code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240118)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240118code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240118)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomoospinadiffexpressionspatiallinearmodelstree40a170898ff370ad9b73f685a331df3fb23002eftargetblank40a1708a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/oospina/diff_expression_spatial_linear_models/tree/40a170898ff370ad9b73f685a331df3fb23002ef" target="_blank">40a1708</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomoospinadiffexpressionspatiallinearmodelstree40a170898ff370ad9b73f685a331df3fb23002eftargetblank40a1708a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/oospina/diff_expression_spatial_linear_models/tree/40a170898ff370ad9b73f685a331df3fb23002ef" target="_blank">40a1708</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/.Rhistory
    Ignored:    code/.DS_Store
    Ignored:    code/diff_expr_hpcrun_spamm/.DS_Store
    Ignored:    code/diff_expr_hpcrun_spamm/code/.DS_Store
    Ignored:    code/diff_expr_hpcrun_spamm/code/.Rhistory
    Ignored:    code/diff_expr_hpcrun_spamm/code/logs/
    Ignored:    code/diff_expr_hpcrun_spamm/data/.DS_Store
    Ignored:    code/diff_expr_hpcrun_spamm/data/geomx_stlist_w_clusters_soabrain.RDS
    Ignored:    code/diff_expr_hpcrun_spamm/data/geomx_stlist_w_clusters_soaliver.RDS
    Ignored:    code/diff_expr_hpcrun_spamm/data/smi_stlist_w_clusters_liver.RDS
    Ignored:    code/diff_expr_hpcrun_spamm/data/smi_stlist_w_clusters_lungcancer.RDS
    Ignored:    code/diff_expr_hpcrun_spamm/data/visium_stlist_w_clusters_maynardbrain.RDS
    Ignored:    code/diff_expr_hpcrun_spamm/data/visium_stlist_w_clusters_raviglioblastoma.RDS
    Ignored:    code/diff_expr_hpcrun_spamm/results/
    Ignored:    code/stdiff_hpcrun_spamm_pairwise_tests/.DS_Store
    Ignored:    code/stdiff_hpcrun_spamm_pairwise_tests/results/
    Ignored:    data/.DS_Store
    Ignored:    data/cosmx_smi_liver_nanostring/
    Ignored:    data/cosmx_smi_lung_nsclc_nanostring/
    Ignored:    data/geomx_spatial_organ_atlas_data/
    Ignored:    data/maynard_2021_prefrontal_cortex/
    Ignored:    data/ravi_2022_glioblastoma/
    Ignored:    data/ravi_gbm_spata_objects/

Unstaged changes:
    Modified:   README.md
    Modified:   analysis/about.Rmd
    Modified:   analysis/diff_expr_example_stdiff.Rmd
    Modified:   analysis/diff_expr_spatial_linear_models.Rmd
    Modified:   analysis/index.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<p>The <code>spatialGE</code> software includes the <code>STdiff</code>
function to test for differentially expressed genes using spatial
covariance structures. The differential expression tests in the <a
href="https://www.nature.com/articles/s41598-024-61758-0">manuscript</a>
associated with this website were performed in an HPC environment.
Nonetheless, the STdiff algorithm can be run on a laptop computer.</p>
<p>Here an example to run the <code>STdiff</code> function is presented
on a small subset of genes to reduce computational time. This vignette
assumes that an <em>STlist</em> object has been already created (click
<a
href="https://fridleylab.github.io/spatialGE/articles/basic_functions_vignette.html">here</a>
for a tutorial to create an <em>STlist</em> object). The <em>STlist</em>
used here will be the same as in the main vignette generating the
figures for the manuscript.</p>
<p>Users are also encouraged to take a look at the <a
href="https://fridleylab.github.io/spatialGE/articles/spatial_differential_expression.html">vignette
in the <code>spatialGE</code> package</a>.</p>
<p>Load libraries</p>
<pre class="r"><code>library(&#39;spatialGE&#39;)
library(&#39;tidyverse&#39;)</code></pre>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.0     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
<p>Load <em>STlist</em> containing the CosMx-SMI lung cancer data set
used in the manuscript.</p>
<pre class="r"><code>stlist_obj = readRDS(&#39;code/diff_expr_hpcrun_spamm/data/smi_stlist_w_clusters_lungcancer.RDS&#39;)</code></pre>
<p>In this small example, the <code>STdiff</code> fuction will be used
to test for differentially expressed genes in tumor cells (“tumor_5”)
and epithelial cells in the FOV “lung5rep3_fov28”. Only the 100 top most
variable will be tested using <em>non-spatial models</em>. Then, 10% of
the differentially expressed genes (i.e., p &lt; 0.05), will be tested
using <em>spatial models</em>.</p>
<p>Counts were already transfomed with the <code>transform_data</code>
function. The <code>STdiff</code> function can be directly called on the
<em>STlist</em> object:</p>
<pre class="r"><code>degs = STdiff(stlist_obj, # STlist object
              annot=&#39;annots&#39;, # Name of the column in stlist_obj@spatial_meta[[&#39;lung5rep3_fov28&#39;]] containing cell types
              samples=&#39;lung5rep3_fov28&#39;, # Name of the sample (i.e., FOV)
              topgenes=100, # Top variable genes to run non-spatial models
              sp_topgenes=0.1, # THe percentage of non-spatial DE tests to re-run using spatial models
              pairwise=T, # Do pairwise comparisons
              clusters=c(&#39;epithelial&#39;, &#39;tumor_5&#39;)) # Specific cell types to test</code></pre>
<pre><code>Testing metadata: annots...
    Running non-spatial mixed models...</code></pre>
<pre><code>Registered S3 methods overwritten by &#39;registry&#39;:
  method               from 
  print.registry_field proxy
  print.registry_entry proxy</code></pre>
<pre><code>    Completed non-spatial mixed models (0.17 min).
    Running spatial tests...</code></pre>
<pre><code>Using paralellisation might be useful. See help(&quot;setNBThreads&quot;)</code></pre>
<pre><code>Using paralellisation might be useful. See help(&quot;setNBThreads&quot;)
Using paralellisation might be useful. See help(&quot;setNBThreads&quot;)
Using paralellisation might be useful. See help(&quot;setNBThreads&quot;)
Using paralellisation might be useful. See help(&quot;setNBThreads&quot;)
Using paralellisation might be useful. See help(&quot;setNBThreads&quot;)</code></pre>
<pre><code>
    Completed spatial mixed models (5.39 min).
STdiff completed in 5.66 min.</code></pre>
<p>The “exp_adj_p_val” column contains the FDR p-values from the
<em>spatial tests</em>. The “adj_p_val” contains the FDR p-values from
the <em>non-spatial tests</em>. The results cna be observed like so:</p>
<pre class="r"><code>degs[[&#39;lung5rep3_fov28&#39;]] %&gt;%
  filter(!is.na(exp_p_val)) # Show only genes with spatial tests </code></pre>
<pre><code># A tibble: 6 × 10
  sample       gene  avg_log2fc cluster_1 cluster_2 mm_p_val adj_p_val exp_p_val
  &lt;chr&gt;        &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 lung5rep3_f… OLFM4       4.62 tumor_5   epitheli…        0         0  0       
2 lung5rep3_f… CEAC…       3.36 tumor_5   epitheli…        0         0  0       
3 lung5rep3_f… CXCL5       2.51 tumor_5   epitheli…        0         0  0       
4 lung5rep3_f… KRT19       2.20 tumor_5   epitheli…        0         0  0       
5 lung5rep3_f… KRT17       2.16 tumor_5   epitheli…        0         0  0       
6 lung5rep3_f… DMBT1       3.03 tumor_5   epitheli…        0         0  1.01e-13
# ℹ 2 more variables: exp_adj_p_val &lt;dbl&gt;, comments &lt;chr&gt;</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.2     readr_2.1.4     tidyr_1.3.0     tibble_3.2.1   
 [9] ggplot2_3.5.0   tidyverse_2.0.0 spatialGE_1.2.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] dotCall64_1.1-1     gtable_0.3.4        spam_2.10-0        
 [4] xfun_0.41           bslib_0.6.1         ggpolypath_0.3.0   
 [7] processx_3.8.3      lattice_0.21-9      numDeriv_2016.8-1.1
[10] callr_3.7.3         tzdb_0.4.0          vctrs_0.6.5        
[13] tools_4.3.2         ps_1.7.5            generics_0.1.3     
[16] parallel_4.3.2      proxy_0.4-27        fansi_1.0.6        
[19] pkgconfig_2.0.3     Matrix_1.6-4        checkmate_2.3.1    
[22] lifecycle_1.0.4     compiler_4.3.2      git2r_0.33.0       
[25] fields_15.2         munsell_0.5.0       getPass_0.2-4      
[28] httpuv_1.6.13       htmltools_0.5.7     maps_3.4.1.1       
[31] sass_0.4.8          yaml_2.3.8          nloptr_2.0.3       
[34] crayon_1.5.2        later_1.3.2         pillar_1.9.0       
[37] jquerylib_0.1.4     whisker_0.4.1       MASS_7.3-60        
[40] cachem_1.0.8        boot_1.3-28.1       nlme_3.1-163       
[43] spaMM_4.4.0         tidyselect_1.2.0    digest_0.6.33      
[46] slam_0.1-50         stringi_1.8.3       rprojroot_2.0.4    
[49] fastmap_1.1.1       grid_4.3.2          colorspace_2.1-0   
[52] cli_3.6.2           magrittr_2.0.3      utf8_1.2.4         
[55] withr_2.5.2         backports_1.4.1     scales_1.3.0       
[58] promises_1.2.1      registry_0.5-1      timechange_0.2.0   
[61] rmarkdown_2.25      httr_1.4.7          hms_1.1.3          
[64] ROI_1.0-1           pbapply_1.7-2       evaluate_0.23      
[67] knitr_1.45          viridisLite_0.4.2   rlang_1.1.2        
[70] Rcpp_1.0.11         glue_1.6.2          minqa_1.2.6        
[73] rstudioapi_0.15.0   jsonlite_1.8.8      R6_2.5.1           
[76] fs_1.6.3           </code></pre>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
